File: src/validate.py
#!/usr/bin/env python3
"""Check every config in this repo against the rules in AGENTS.md.

Exits 0 if everything passes, 1 if anything fails. Every failure prints the
file, the line, and what to do about it.

Usage:
    python3 src/validate.py                 # all checks (downloads arg.cpp once)
    python3 src/validate.py --skip-keys     # offline: skip the llama.cpp key check
    python3 src/validate.py --arg-cpp PATH  # use a local copy of arg.cpp
"""

import argparse
import json
import pathlib
import re
import sys
import urllib.request

ROOT = pathlib.Path(__file__).resolve().parent.parent
CONFIGS = ROOT / "configs"
DIST = ROOT / "dist"
ESTIMATES = CONFIGS / "vram-estimates.json"

ARG_CPP_URL = "https://raw.githubusercontent.com/ggml-org/llama.cpp/master/common/arg.cpp"
CACHE = pathlib.Path("/tmp/llama-cpp-arg.cpp")

MIN_CTX = 4096
FORBIDDEN_KEYS = {"host", "port", "api-key"}
# KV cache types the profiles in build.py need a measured curve for.
NEEDED_KV = {"f16", "q8_0", "q5_1", "q4_0"}
ALLOWED_QUANTS = {
    "UD-Q3_K_XL", "Q4_K_M", "UD-Q4_K_M", "UD-Q4_K_XL",
    "Q5_K_M", "UD-Q5_K_M", "UD-Q5_K_XL", "Q6_K", "UD-Q6_K",
}

failures = []


def fail(where, msg, fix):
    failures.append(f"{where}\n    problem: {msg}\n    fix:     {fix}")


def parse_ini(path):
    """Yield (line_number, section, key, value). Section is None before the first header."""
    section = None
    for n, raw in enumerate(path.read_text().splitlines(), 1):
        line = raw.strip()
        if not line or line.startswith((";", "#")):
            continue
        if line.startswith("["):
            section = line.strip("[]")
            yield n, section, None, None
            continue
        if "=" in line:
            key, value = line.split("=", 1)
            yield n, section, key.strip(), value.strip()


def load_valid_keys(args):
    if args.skip_keys:
        return None
    if args.arg_cpp:
        text = pathlib.Path(args.arg_cpp).read_text()
    elif CACHE.exists():
        text = CACHE.read_text()
    else:
        try:
            text = urllib.request.urlopen(ARG_CPP_URL, timeout=60).read().decode()
            CACHE.write_text(text)
        except Exception as e:
            print(f"could not download arg.cpp ({e}); re-run with --skip-keys", file=sys.stderr)
            sys.exit(2)

    keys = {k.lstrip("-") for k in re.findall(r'"(--?[A-Za-z0-9][A-Za-z0-9._-]*)"', text)}
    keys |= set(re.findall(r'set_env\("([A-Z0-9_]+)"\)', text))
    marker = "void common_params_add_preset_options"
    if marker in text:
        keys |= set(re.findall(r'\{"([a-z][a-z0-9-]*)"\}', text[text.index(marker):]))
    return keys


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--skip-keys", action="store_true", help="skip the llama.cpp key check")
    ap.add_argument("--arg-cpp", help="path to a local common/arg.cpp")
    args = ap.parse_args()

    valid_keys = load_valid_keys(args)
    estimates = json.loads(ESTIMATES.read_text())["models"]
    configs = sorted(CONFIGS.rglob("*.ini"))

    if not configs:
        fail("configs/", "no config files found", "add configs/<provider>/<model>.ini")

    seen_ids = {}

    for path in configs:
        rel = path.relative_to(ROOT)
        sections = [s for _, s, k, _ in parse_ini(path) if k is None]

        if len(sections) != 1:
            fail(f"{rel}", f"expected exactly 1 section, found {len(sections)}",
                 "one model per file; build.py writes the [*] section itself")
            continue

        model_id = sections[0]

        if model_id == "*":
            fail(f"{rel}", "contains a [*] section",
                 "remove it; build.py writes [*] into dist/ files")
        if model_id != model_id.lower():
            fail(f"{rel}", f"model id '{model_id}' is not lowercase",
                 f"rename the section and file to '{model_id.lower()}'")
        if path.stem != model_id:
            fail(f"{rel}", f"file name '{path.stem}' does not match section '{model_id}'",
                 f"rename the file to '{model_id}.ini' (use git mv)")
        if model_id in seen_ids:
            fail(f"{rel}", f"model id '{model_id}' also defined in {seen_ids[model_id]}",
                 "model ids must be unique")
        seen_ids[model_id] = rel
        if model_id not in estimates:
            fail(f"{rel}", f"no VRAM data for '{model_id}' in configs/vram-estimates.json",
                 "measure it with gguf-parser and add it; see AGENTS.md")

        for n, _, key, value in parse_ini(path):
            if key is None:
                continue
            if key in FORBIDDEN_KEYS:
                fail(f"{rel}:{n}", f"'{key}' does nothing in a config file",
                     f"remove it and pass --{key} on the llama-server command line")
            if valid_keys is not None and key not in valid_keys:
                fail(f"{rel}:{n}", f"llama.cpp does not accept the key '{key}'",
                     "the server would exit with code 1; check the spelling against arg.cpp")
            if ";" in value or "#" in value:
                fail(f"{rel}:{n}", f"value for '{key}' contains ';' or '#'",
                     "llama.cpp silently deletes the rest of the line; remove the character")
            if value.startswith("/") or value.startswith("~") or "{" in key:
                fail(f"{rel}:{n}", f"'{key}' looks like a file path",
                     "configs must not contain paths; models come from --models-dir")
            if key in ("model", "mmproj", "model-draft", "hf-repo"):
                fail(f"{rel}:{n}", f"'{key}' hardcodes a model location",
                     "remove it; llama.cpp finds the files via --models-dir")
            if key == "ctx-size":
                # Optional. Present means "never give this model more than N",
                # which build.py respects; absent means "as much as fits".
                if not value.isdigit() or int(value) < MIN_CTX:
                    fail(f"{rel}:{n}", f"ctx-size cap '{value}' is not a number >= {MIN_CTX}",
                         "remove it to let build.py choose, or set a sensible cap")

    where = "configs/vram-estimates.json"
    for model_id, data in estimates.items():
        if model_id not in seen_ids:
            fail(where, f"'{model_id}' has VRAM data but no config file",
                 f"add configs/<provider>/{model_id}.ini or remove the entry")
        for quant in data.get("quants", {}):
            if quant not in ALLOWED_QUANTS:
                fail(where, f"'{model_id}' uses quant '{quant}'",
                     f"allowed quants: {', '.join(sorted(ALLOWED_QUANTS))}")
        if not data.get("quants"):
            fail(where, f"'{model_id}' has no quant measurements",
                 f"run: python3 src/measure.py --quants {model_id}")
        curves = data.get("ref_curves", {})
        missing_kv = sorted(NEEDED_KV - set(curves))
        if not curves or not data.get("max_ctx"):
            fail(where, f"'{model_id}' has no context curves",
                 f"run: python3 src/measure.py --context {model_id}")
        elif missing_kv:
            fail(where, f"'{model_id}' has no curve for KV type(s): {', '.join(missing_kv)}",
                 f"run: python3 src/measure.py --context {model_id}")
        elif data.get("ref_quant") not in data.get("quants", {}):
            fail(where, f"'{model_id}' ref_quant '{data.get('ref_quant')}' is missing from quants",
                 "the context curves cannot be offset without it; re-measure the model")
        else:
            for kv, curve in curves.items():
                if str(data["max_ctx"]) not in curve:
                    fail(where, f"'{model_id}' {kv} curve has no point at max_ctx {data['max_ctx']}",
                         f"run: python3 src/measure.py --context {model_id}")

    if not DIST.exists() or not list(DIST.glob("*.ini")):
        fail("dist/", "no generated files found", "run: python3 src/build.py")

    if failures:
        print(f"FAILED - {len(failures)} problem(s)\n")
        for f in failures:
            print(f"  {f}\n")
        return 1

    print(f"OK - {len(configs)} configs, {len(estimates)} models, "
          f"{len(list(DIST.glob('*.ini')))} generated files")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/internal/preset/validate.go
package preset

import (
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"
)

// Checks every config in this repo against the rules in AGENTS.md.
// Every failure prints the file, the line, and what to do about it.

const (
	argCppURL   = "https://raw.githubusercontent.com/ggml-org/llama.cpp/master/common/arg.cpp"
	argCppCache = "/tmp/llama-cpp-arg.cpp"
)

var (
	forbiddenKeys = []string{"host", "port", "api-key"}
	// KV cache types the profiles in plan.go need a measured curve for.
	neededKV = []string{"f16", "q8_0", "q5_1", "q4_0"}
	// The band is QuantLadder in plan.go.
	allowedQuants = QuantLadder
	pathLikeKeys  = []string{"model", "mmproj", "model-draft", "hf-repo"}
)

type ValidateOptions struct {
	SkipKeys bool
	ArgCpp   string
}

type failure struct{ where, problem, fix string }

type checker struct {
	root     string
	failures []failure
}

func (c *checker) fail(where, problem, fix string) {
	c.failures = append(c.failures, failure{where, problem, fix})
}

func Validate(root string, opts ValidateOptions) int {
	c := &checker{root: root}

	validKeys, err := loadValidKeys(opts)
	if err != nil {
		fmt.Fprintf(os.Stderr, "%v; re-run with --skip-keys\n", err)
		return 2
	}

	est, err := LoadEstimates(root)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 2
	}
	configs, err := LoadModelConfigs(root)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 2
	}
	if len(configs) == 0 {
		c.fail("configs/", "no config files found", "add configs/<provider>/<model>.ini")
	}

	c.checkConfigs(configs, est, validKeys)
	c.checkEstimates(configs, est)
	c.checkDist()

	if len(c.failures) > 0 {
		fmt.Printf("FAILED - %d problem(s)\n\n", len(c.failures))
		for _, f := range c.failures {
			fmt.Printf("  %s\n    problem: %s\n    fix:     %s\n\n", f.where, f.problem, f.fix)
		}
		return 1
	}
	fmt.Printf("OK - %d configs, %d models, %d generated files\n",
		len(configs), len(est.IDs), len(distFiles(root)))
	return 0
}

func (c *checker) checkConfigs(configs []*ModelConfig, est *Estimates, validKeys map[string]bool) {
	seen := map[string]string{}

	for _, cfg := range configs {
		raw, err := os.ReadFile(filepath.Join(c.root, cfg.Path))
		if err != nil {
			c.fail(cfg.Path, err.Error(), "make the file readable")
			continue
		}
		text := string(raw)

		var sections []string
		for _, e := range ParseIni(text) {
			if e.Key == "" {
				sections = append(sections, e.Section)
			}
		}
		if len(sections) != 1 {
			c.fail(cfg.Path, fmt.Sprintf("expected exactly 1 section, found %d", len(sections)),
				"one model per file; build writes the [*] section itself")
			continue
		}
		id := sections[0]

		if id == "*" {
			c.fail(cfg.Path, "contains a [*] section", "remove it; build writes [*] into dist/ files")
		}
		if id != strings.ToLower(id) {
			c.fail(cfg.Path, fmt.Sprintf("model id '%s' is not lowercase", id),
				fmt.Sprintf("rename the section and file to '%s'", strings.ToLower(id)))
		}
		stem := strings.TrimSuffix(filepath.Base(cfg.Path), ".ini")
		if stem != id {
			c.fail(cfg.Path, fmt.Sprintf("file name '%s' does not match section '%s'", stem, id),
				fmt.Sprintf("rename the file to '%s.ini' (use git mv)", id))
		}
		if prev, ok := seen[id]; ok {
			c.fail(cfg.Path, fmt.Sprintf("model id '%s' also defined in %s", id, prev),
				"model ids must be unique")
		}
		seen[id] = cfg.Path
		if _, ok := est.Data[id]; !ok {
			c.fail(cfg.Path, fmt.Sprintf("no VRAM data for '%s' in configs/vram-estimates.json", id),
				"run: make measure")
		}

		head := text
		if i := strings.Index(text, "["+id+"]"); i >= 0 {
			head = text[:i]
		}
		for _, f := range []struct{ field, example string }{
			{"repo", "unsloth/Qwen3-8B-GGUF"}, {"params", "8B"},
		} {
			if headField(head, f.field) == "" {
				c.fail(cfg.Path, fmt.Sprintf("no '; %s:' comment", f.field),
					fmt.Sprintf("add a line like '; %s: %s'; MODELS.md is built from it", f.field, f.example))
			}
		}

		for _, e := range ParseIni(text) {
			if e.Key == "" {
				continue
			}
			at := fmt.Sprintf("%s:%d", cfg.Path, e.Line)
			if contains(forbiddenKeys, e.Key) {
				c.fail(at, fmt.Sprintf("'%s' does nothing in a config file", e.Key),
					fmt.Sprintf("remove it and pass --%s on the llama-server command line", e.Key))
			}
			if validKeys != nil && !validKeys[e.Key] {
				c.fail(at, fmt.Sprintf("llama.cpp does not accept the key '%s'", e.Key),
					"the server would exit with code 1; check the spelling against arg.cpp")
			}
			if strings.ContainsAny(e.Value, ";#") {
				c.fail(at, fmt.Sprintf("value for '%s' contains ';' or '#'", e.Key),
					"llama.cpp silently deletes the rest of the line; remove the character")
			}
			if strings.HasPrefix(e.Value, "/") || strings.HasPrefix(e.Value, "~") || strings.Contains(e.Key, "{") {
				c.fail(at, fmt.Sprintf("'%s' looks like a file path", e.Key),
					"configs must not contain paths; models come from --models-dir")
			}
			if contains(pathLikeKeys, e.Key) {
				c.fail(at, fmt.Sprintf("'%s' hardcodes a model location", e.Key),
					"remove it; llama.cpp finds the files via --models-dir")
			}
			if e.Key == "ctx-size" {
				// Optional. Present means "never give this model more than N",
				// which build respects; absent means "as much as fits".
				n, err := strconv.Atoi(e.Value)
				if err != nil || n < MinCtx {
					c.fail(at, fmt.Sprintf("ctx-size cap '%s' is not a number >= %d", e.Value, MinCtx),
						"remove it to let build choose, or set a sensible cap")
				}
			}
		}
	}
}

func (c *checker) checkEstimates(configs []*ModelConfig, est *Estimates) {
	const where = "configs/vram-estimates.json"

	haveConfig := map[string]bool{}
	for _, cfg := range configs {
		haveConfig[cfg.ID] = true
	}

	for _, id := range est.IDs {
		d := est.Data[id]
		if !haveConfig[id] {
			c.fail(where, fmt.Sprintf("'%s' has VRAM data but no config file", id),
				fmt.Sprintf("add configs/<provider>/%s.ini or remove the entry", id))
		}
		for _, quant := range d.QuantKeys {
			if !contains(allowedQuants, quant) {
				sorted := append([]string(nil), allowedQuants...)
				sort.Strings(sorted)
				c.fail(where, fmt.Sprintf("'%s' uses quant '%s'", id, quant),
					"allowed quants: "+strings.Join(sorted, ", "))
			}
		}
		if len(d.QuantKeys) == 0 {
			c.fail(where, fmt.Sprintf("'%s' has no quant measurements", id),
				"run: make measure MODEL="+id)
		}

		var missingKV []string
		for _, kv := range neededKV {
			if _, ok := d.Curves[kv]; !ok {
				missingKV = append(missingKV, kv)
			}
		}
		sort.Strings(missingKV)

		switch {
		case len(d.Curves) == 0 || d.MaxCtx == 0:
			c.fail(where, fmt.Sprintf("'%s' has no context curves", id),
				"run: make measure MODEL="+id)
		case len(missingKV) > 0:
			c.fail(where, fmt.Sprintf("'%s' has no curve for KV type(s): %s", id, strings.Join(missingKV, ", ")),
				"run: make measure MODEL="+id)
		case !contains(d.QuantKeys, d.RefQuant):
			c.fail(where, fmt.Sprintf("'%s' ref_quant '%s' is missing from quants", id, d.RefQuant),
				"the context curves cannot be offset without it; re-measure the model")
		default:
			for _, kv := range d.KVKeys {
				if _, ok := d.Curves[kv].Vals[strconv.Itoa(d.MaxCtx)]; !ok {
					c.fail(where, fmt.Sprintf("'%s' %s curve has no point at max_ctx %d", id, kv, d.MaxCtx),
						"run: make measure MODEL="+id)
				}
			}
		}
	}
}

func (c *checker) checkDist() {
	files := distFiles(c.root)
	if len(files) == 0 {
		c.fail("dist/", "no generated files found", "run: make build")
		return
	}
	// A repeated section name makes llama.cpp refuse to start with
	// "model 'x' appears multiple times", so catch it before a release does.
	for _, path := range files {
		raw, err := os.ReadFile(path)
		if err != nil {
			continue
		}
		count := map[string]int{}
		for _, e := range ParseIni(string(raw)) {
			if e.Key == "" && e.Section != "*" {
				count[e.Section]++
			}
		}
		var dupes []string
		for name, n := range count {
			if n > 1 {
				dupes = append(dupes, name)
			}
		}
		sort.Strings(dupes)
		for _, d := range dupes {
			c.fail(rel(c.root, path), fmt.Sprintf("section '%s' appears more than once", d),
				"two setups produced the same name; fix CtxSlug() in src/internal/preset/plan.go")
		}
	}
}

func distFiles(root string) []string {
	files, _ := filepath.Glob(filepath.Join(root, "dist", "*.ini"))
	sort.Strings(files)
	return files
}

var (
	flagRe   = regexp.MustCompile(`"(--?[A-Za-z0-9][A-Za-z0-9._-]*)"`)
	envRe    = regexp.MustCompile(`set_env\("([A-Z0-9_]+)"\)`)
	presetRe = regexp.MustCompile(`\{"([a-z][a-z0-9-]*)"\}`)
)

// loadValidKeys returns the set of keys llama.cpp accepts, or nil when the
// check is skipped.
func loadValidKeys(opts ValidateOptions) (map[string]bool, error) {
	if opts.SkipKeys {
		return nil, nil
	}
	var text string
	switch {
	case opts.ArgCpp != "":
		raw, err := os.ReadFile(opts.ArgCpp)
		if err != nil {
			return nil, err
		}
		text = string(raw)
	default:
		if raw, err := os.ReadFile(argCppCache); err == nil {
			text = string(raw)
			break
		}
		client := &http.Client{Timeout: 60 * time.Second}
		resp, err := client.Get(argCppURL)
		if err != nil {
			return nil, fmt.Errorf("could not download arg.cpp (%w)", err)
		}
		defer func() { _ = resp.Body.Close() }() // nothing useful to do with a close error
		if resp.StatusCode != http.StatusOK {
			return nil, fmt.Errorf("could not download arg.cpp (status %s)", resp.Status)
		}
		raw, err := io.ReadAll(resp.Body)
		if err != nil {
			return nil, fmt.Errorf("could not download arg.cpp (%w)", err)
		}
		text = string(raw)
		_ = os.WriteFile(argCppCache, raw, 0o644)
	}

	keys := map[string]bool{}
	for _, m := range flagRe.FindAllStringSubmatch(text, -1) {
		keys[strings.TrimLeft(m[1], "-")] = true
	}
	for _, m := range envRe.FindAllStringSubmatch(text, -1) {
		keys[m[1]] = true
	}
	// Some keys exist only in config files, not on the command line.
	const marker = "void common_params_add_preset_options"
	if i := strings.Index(text, marker); i >= 0 {
		for _, m := range presetRe.FindAllStringSubmatch(text[i:], -1) {
			keys[m[1]] = true
		}
	}
	return keys, nil
}

func contains(haystack []string, needle string) bool {
	for _, s := range haystack {
		if s == needle {
			return true
		}
	}
	return false
}
